File: r/03.R
library(tidyverse)

input <- read_delim("raw/03.txt", delim = ",", col_names = FALSE)


# Part One ----------------------------------------------------------------

tob_positions <- input %>% 
  rename(tree_map = X1) %>% 
  mutate(x_pos = 1 + row_number() * 3 - 3,
         map_len = str_length(tree_map)) %>% 
  mutate(x_pos_mod = x_pos %% map_len,
         x_pos_mod = ifelse(x_pos_mod == 0, map_len, x_pos_mod)) %>% 
  mutate(pos = substr(tree_map, x_pos_mod, x_pos_mod))

sum(tob_positions$pos == "#")


# Part Two ----------------------------------------------------------------

# Right 1, down 1.
# Right 3, down 1. (This is the slope you already checked.)
# Right 5, down 1.
# Right 7, down 1.

tree_hits <- function(n_right) {
  tob_positions <- input %>% 
    rename(tree_map = X1) %>% 
    mutate(x_pos = 1 + row_number() * n_right - n_right,
           map_len = str_length(tree_map)) %>% 
    mutate(x_pos_mod = x_pos %% map_len,
           x_pos_mod = ifelse(x_pos_mod == 0, map_len, x_pos_mod)) %>% 
    mutate(pos = substr(tree_map, x_pos_mod, x_pos_mod))
  
  sum(tob_positions$pos == "#")
}

slopes <- c(1, 3, 5, 7)

hits <- map_dbl(slopes, tree_hits)

# Right 1, down 2.

n_right <- 1

tob_positions_odds <- input %>% 
  rename(tree_map = X1) %>% 
  filter(row_number() %% 2 == 1) %>% 
  mutate(x_pos = 1 + row_number() * n_right - n_right,
         map_len = str_length(tree_map)) %>% 
  mutate(x_pos_mod = x_pos %% map_len,
         x_pos_mod = ifelse(x_pos_mod == 0, map_len, x_pos_mod)) %>% 
  mutate(pos = substr(tree_map, x_pos_mod, x_pos_mod))

hits_odd <- sum(tob_positions_odds$pos == "#")

reduce(c(hits, hits_odd), `*`)

File: py/03.py
file = open("raw/03.txt")

input = file.read()
input_lst = input.split("\n")
input_lst

## Part One

rows = len(input_lst)
cols = len(input_lst[0])
pos = list(range(0, rows - 1))
pos = [i * 3 % cols for i in pos]

input_zipped = list(zip(input_lst, pos))
input_zipped[1][1]
sum([i[0][i[1]] == "#" for i in input_zipped])

## Part Two

# Right 1, down 1.
# Right 3, down 1. (This is the slope you already checked.)
# Right 5, down 1.
# Right 7, down 1.
# Right 1, down 2.


def count_trees(input_map, right, down):
    rows = len(input_map)
    cols = len(input_map[0])
    row_range = list(range(0, rows - 1))
    pos_row = [i % down == 0 for i in row_range]
    filtered_input_map = [i[0] for i in list(zip(input_map, pos_row)) if i[1]]

    row_range_filtered = list(range(0, len(filtered_input_map)))
    pos = [i * right % cols for i in row_range_filtered]

    input_zipped = list(zip(filtered_input_map, pos))
    return sum([i[0][i[1]] == "#" for i in input_zipped])

count_trees(input_lst, 1, 2)

count_trees(input_lst, 1, 1) * count_trees(input_lst, 3, 1) * count_trees(input_lst, 5, 1) * count_trees(input_lst, 7, 1) * count_trees(input_lst, 1, 2)
